module/zfs/dnode.c

```diff
@@ -1326,7 +1326,15 @@ dnode_hold_impl(objset_t *os, uint64_t object, int flag, int slots,
 				continue;
 			}
 
-			dn = dnode_create(os, dn_block + idx, db, object, dnh);
+			/*
+			 * Someone else won the race and called dnode_create()
+			 * after we checked DN_SLOT_IS_PTR() above but before
+			 * we acquired the lock.
+			 */
+			if (DN_SLOT_IS_PTR(dnh->dnh_dnode))
+				dn = dnh->dnh_dnode;
```

Comment:
What do you think about adding a `dnode_hold_alloc_lock_misses` kstat here as a counterpart to `dnode_hold_free_lock_misses`.  It could be insightful to know how often this actually happens.

Reply:
I'll do that.